pkg/rtc/uptrackmanager.go

```diff
@@ -0,0 +1,676 @@
+		u.twcc.OnFeedback(func(pkt rtcp.RawPacket) {
+			if u.onWriteRTCP != nil {
+				u.onWriteRTCP([]rtcp.Packet{&pkt})
+			}
+		})
+	}
+	u.lock.Unlock()
+
+	mt.AddReceiver(rtpReceiver, track, u.twcc)
+
+	if newTrack {
+		u.handleTrackPublished(mt)
+	}
+}
+
+// should be called with lock held
+func (u *UptrackManager) getPublishedTrack(sid string) types.PublishedTrack {
+	return u.publishedTracks[sid]
+}
+
+// should be called with lock held
+func (u *UptrackManager) getPublishedTrackBySignalCid(clientId string) types.PublishedTrack {
+	for _, publishedTrack := range u.publishedTracks {
+		if publishedTrack.SignalCid() == clientId {
+			return publishedTrack
+		}
+	}
+
+	return nil
+}
+
+// should be called with lock held
+func (u *UptrackManager) getPublishedTrackBySdpCid(clientId string) types.PublishedTrack {
+	for _, publishedTrack := range u.publishedTracks {
+		if publishedTrack.SdpCid() == clientId {
+			return publishedTrack
+		}
+	}
+
+	return nil
+}
+
+// should be called with lock held
+func (u *UptrackManager) getPendingTrack(clientId string, kind livekit.TrackType) (string, *livekit.TrackInfo) {
+	signalCid := clientId
+	trackInfo := u.pendingTracks[clientId]
+
+	if trackInfo == nil {
+		//
+		// If no match on client id, find first one matching type
+		// as MediaStreamTrack can change client id when transceiver
+		// is added to peer connection.
+		//
+		for cid, ti := range u.pendingTracks {
+			if ti.Type == kind {
+				trackInfo = ti
+				signalCid = cid
+				break
+			}
+		}
+	}
+
+	// if still not found, we are done
+	if trackInfo == nil {
+		u.params.Logger.Errorw("track info not published prior to track", nil, "clientId", clientId)
+	}
+	return signalCid, trackInfo
+}
+
+func (u *UptrackManager) handleTrackPublished(track types.PublishedTrack) {
+	u.lock.Lock()
+	if _, ok := u.publishedTracks[track.ID()]; !ok {
+		u.publishedTracks[track.ID()] = track
+	}
+	u.lock.Unlock()
+
+	track.AddOnClose(func() {
+		// cleanup
+		u.lock.Lock()
+		trackSid := track.ID()
+		delete(u.publishedTracks, trackSid)
+		delete(u.pendingSubscriptions, trackSid)
+		// not modifying subscription permissions, will get reset on next update from participant
+		u.lock.Unlock()
+		// only send this when client is in a ready state
+		if u.onTrackUpdated != nil {
+			u.onTrackUpdated(track, true)
+		}
+	})
+
+	if u.onTrackPublished != nil {
+		u.onTrackPublished(track)
+	}
+}
+
+func (u *UptrackManager) updateSubscriptionPermissions(permissions *livekit.UpdateSubscriptionPermissions) {
+	// every update overrides the existing
+
+	// all_participants takes precedence
+	if permissions.AllParticipants {
+		// everything is allowed, nothing else to do
+		u.subscriptionPermissions = nil
+		return
+	}
+
+	u.subscriptionPermissions = make(map[string]*livekit.TrackPermission)
+
+	// if track_permissions is empty, nobody is allowed to subscribe to any tracks
+	if permissions.TrackPermissions != nil && len(permissions.TrackPermissions) == 0 {
```

Comment:
FYI: len check will return 0 for nil slices.

Reply:
yeah, forgot to remove it. I guess I can remove those line of code completely and let the range take care of it. Will do that before merging.